2.8.0 staging

alexlamsl opened the pull request:
This is a test merge of all 22 open PRs mentioned in #1383 - list of fix-ups:
- #1454 - update tests
- #1451 - conflicts with #1457
- #1461 - conflicts with #1451
- #1476 - conflicts with #1450
- #1477 - conflicts with #1451
- #1469 - conflicts with #1473
- #1483 - conflicts with #1469
- #1487 - conflicts with #1461
- #1494 - conflicts with #1448 & #1451

/cc @avdg @kzc @rvanvelzen

alexlamsl commented:
Not counting `README.md` and tests, these commits added 206 lines in total.

(Travis CI [timed-out](https://travis-ci.org/mishoo/UglifyJS2/jobs/200607597) [here](https://github.com/mishoo/UglifyJS2/blob/642273c29002e7676719f489d7fcf552974118f4/test/mocha/number-literal.js#L5). Not sure if anything should/can be done about it?)

avdg commented:
Yeah, I should increase the timeout in #1366

avdg commented:
Updated timeout to `15000` instead of `5000` in #1366

alexlamsl commented:
@avdg I'm confused - how would increasing time-out period on `test/mocha/portable.js` fixes a time-out issue on `test/mocha/number-literal.js`?

alexlamsl commented:
Tested the follow options on `test/jetstream.js`:
```
-mc warnings=false
-m
-b
-mc passes=2,warnings=false
-mc passes=3,warnings=false
-mc unsafe,warnings=false
-mc unsafe,passes=2,warnings=false
-mc unsafe,passes=3,warnings=false
-mc collapse_vars,warnings=false
-mc collapse_vars,passes=2,warnings=false
-mc collapse_vars,passes=3,warnings=false
-mc reduce_vars,warnings=false
-mc reduce_vars,passes=2,warnings=false
-mc reduce_vars,passes=3,warnings=false
-mc collapse_vars,reduce_vars,warnings=false
-mc collapse_vars,reduce_vars,passes=2,warnings=false
-mc collapse_vars,reduce_vars,passes=3,warnings=false
-mc unsafe,collapse_vars,reduce_vars,warnings=false
-mc unsafe,collapse_vars,reduce_vars,passes=2,warnings=false
-mc unsafe,collapse_vars,reduce_vars,passes=3,warnings=false
-mc unsafe,keep_fargs=0,collapse_vars,reduce_vars,warnings=false
-mc unsafe,keep_fargs=0,collapse_vars,reduce_vars,passes=2,warnings=false
-mc unsafe,keep_fargs=0,collapse_vars,reduce_vars,passes=3,warnings=false
```

alexlamsl commented:
`test/benchmark.js`

### [2.7.5](https://github.com/mishoo/UglifyJS2/tree/v2.7.5)

```
https://code.jquery.com/jquery-3.1.1.js
Timing information (compressed 1 files):
- parse: 0.197s
- scope: 0.288s
- squeeze: 0.770s
- mangle: 0.031s
- generate: 0.100s

SHA1: 441e618ce79ed77d33e94a5ea8339f503d3cf892

https://code.angularjs.org/1.6.1/angular.js
Timing information (compressed 1 files):
- parse: 0.232s
- scope: 0.565s
- squeeze: 0.990s
- mangle: 0.065s
- generate: 0.184s

SHA1: b6d96dfb5a8c75641194c4c597b30a909862ab39

https://cdnjs.cloudflare.com/ajax/libs/mathjs/3.9.0/math.js
Timing information (compressed 1 files):
- parse: 0.510s
- scope: 1.560s
- squeeze: 17.244s
- mangle: 0.145s
- generate: 0.500s

SHA1: b36db92ef7c06e7b130beb7a3b01b02cbcdc2f95

https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.js
Timing information (compressed 1 files):
- parse: 0.062s
- scope: 0.118s
- squeeze: 0.206s
- mangle: 0.013s
- generate: 0.040s

SHA1: da4c56cce3d0a6abd94b4024261a3387af536408

https://unpkg.com/react@15.3.2/dist/react.js
Timing information (compressed 1 files):
- parse: 0.214s
- scope: 0.559s
- squeeze: 1.021s
- mangle: 0.053s
- generate: 0.173s

SHA1: aad6e700c1de39ad1d8310c8fd490d52f69b12c9

http://builds.emberjs.com/tags/v2.11.0/ember.prod.js
Timing information (compressed 1 files):
- parse: 0.501s
- scope: 1.286s
- squeeze: 2.187s
- mangle: 0.143s
- generate: 0.451s

SHA1: 6fe981090b52f6ff878120501d75d0f0fb345dc8

https://cdn.jsdelivr.net/lodash/4.17.4/lodash.js
Timing information (compressed 1 files):
- parse: 0.126s
- scope: 0.378s
- squeeze: 0.636s
- mangle: 0.071s
- generate: 0.108s

SHA1: 2910100ae329aae2b9d6465315cd1d1c4b6601f5

https://cdnjs.cloudflare.com/ajax/libs/d3/4.5.0/d3.js
Timing information (compressed 1 files):
- parse: 0.255s
- scope: 0.839s
- squeeze: 1.357s
- mangle: 0.197s
- generate: 0.273s

SHA1: 7fc8a34569ae383780fce638ef91680ef05fcde6
```

### This PR

```
https://code.jquery.com/jquery-3.1.1.js
Timing information (compressed 1 files):
- parse: 0.127s
- scope: 0.328s
- squeeze: 0.852s
- mangle: 0.033s
- generate: 0.112s

SHA1: 7542a93fe3b60517a4e385c7cc9f14ab7f969904

https://code.angularjs.org/1.6.1/angular.js
Timing information (compressed 1 files):
- parse: 0.246s
- scope: 0.541s
- squeeze: 1.252s
- mangle: 0.065s
- generate: 0.197s

SHA1: 1d560a92dc854e35ad111ca855bc49ba12103cac

https://cdnjs.cloudflare.com/ajax/libs/mathjs/3.9.0/math.js
Timing information (compressed 1 files):
- parse: 0.494s
- scope: 1.452s
- squeeze: 17.854s
- mangle: 0.155s
- generate: 0.493s

SHA1: 09a901226439aa3273875d4e1b848261880ad08a

https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.js
Timing information (compressed 1 files):
- parse: 0.044s
- scope: 0.120s
- squeeze: 0.234s
- mangle: 0.013s
- generate: 0.039s

SHA1: 7490b7e2eb93461408c7d97a083a2ffd897a7809

https://unpkg.com/react@15.3.2/dist/react.js
Timing information (compressed 1 files):
- parse: 0.207s
- scope: 0.514s
- squeeze: 1.237s
- mangle: 0.050s
- generate: 0.186s

SHA1: e34ee5dd004d1db6f2582d2d17d8cee1c430a4b0

http://builds.emberjs.com/tags/v2.11.0/ember.prod.js
Timing information (compressed 1 files):
- parse: 0.507s
- scope: 1.302s
- squeeze: 2.619s
- mangle: 0.149s
- generate: 0.460s

SHA1: d77910641acbc94528e8a7a3ac302170eb7454b8

https://cdn.jsdelivr.net/lodash/4.17.4/lodash.js
Timing information (compressed 1 files):
- parse: 0.148s
- scope: 0.288s
- squeeze: 0.710s
- mangle: 0.057s
- generate: 0.099s

SHA1: 4e24e6f70e574b06afbf5e8654740a80ee97d6b9

https://cdnjs.cloudflare.com/ajax/libs/d3/4.5.0/d3.js
Timing information (compressed 1 files):
- parse: 0.264s
- scope: 0.947s
- squeeze: 1.629s
- mangle: 0.161s
- generate: 0.263s

SHA1: 06439dfb30a7ec469d4d6f250f90a06b32b65657
```

alexlamsl commented:
`uglify-js -mc`

File | Original | 2.7.5 | This PR
--- | ---:| ---:| ---:
angular-1.6.1.js | 1,232,632 | 174,174 | 174,065
boostrap.js | 69,707 | 36,920 | 36,901
d3.js | 451,131 | 214,137 | 214,126
ember.prod.js | 1,852,178 | 530,719 | 530,456
jquery-3.1.1.js | 267,194 | 87,192 | 87,062
lodash.js | 539,590 | 71,380 | 71,377
math.js | 1,590,107 | 470,543 | 470,355
react.js | 701,412 | 207,922 | 207,781

avdg commented:
uh, maybe I didn't do a complete check. The fix was suppose to solve 1 known problem anyway.

avdg commented:
See https://travis-ci.org/mishoo/UglifyJS2/jobs/200607597#L772

alexlamsl commented:
@avdg you are right and I'm just terminally confused 😅 

FWIW, I did a search of `Should not allow legacy octal literals in strict mode` across the code base and `test/mocha/number-literal.js` is the only file that has it. But yes, as you stated it's also under the heading of `portable` which is where the `this.timeout(5000);` is at.

kzc commented:
@alexlamsl AFAIK only #1451 conflicts with the other pending PRs. 

If you changed this PR's title to "2.8.0 staging" then @rvanvelzen could simply merge this for the next release. Mind you, if we did that then the underlying PRs could no longer be updated as changes would not be picked up.

alexlamsl commented:
@kzc may be due you used 3-way merge whereas I did `git merge --squash` which probably didn't.

I was thinking @rvanvelzen would do the merge, then compare notes with this PR so we can both catch any problems if they turn out to be different. Let's just say my git-fu is not at the level where I can confidently say I did everything right. 😅

kzc commented:
Unrelated: I think you could add `pure_getters` and `unsafe_comps` to your list of options to test with jetstream:
```
-mc unsafe,unsafe_comps,pure_getters,keep_fargs=0,collapse_vars,reduce_vars,passes=3,warnings=false
```

alexlamsl commented:
Also unrelated: in manually going over the files, so far I noticed when `reduce_vars` eventually causes a variable to be unused which is being assigned a RegExp as value, we will be left with the value as it is now considered to `.has_side_effects()`.

alexlamsl commented:
@kzc your suggested options (& two `passes` variant) have completed successfully 😉

alexlamsl commented:
https://github.com/mishoo/UglifyJS2/pull/1485#issuecomment-279164291 boils down to this:
```
echo 'function f(){var a=/b/;}' | uglifyjs -c
```
2.7.5 gives:
```
WARN: Dropping unused variable a [-:1,17]
function f(){}
```
This PR gives:
```
WARN: Side effects in initialization of unused variable a [-:1,17]
function f(){/b/}
```

kzc commented:
RegExp literals are tricky. Unlike strings they are not truly constant - they are mutable and hold state. I'd rather err on the side of caution here rather than risk generating bad code.

kzc commented:
```js
        (function(){
            var result, rx = /ab*/g;
            while (result = rx.exec('acdabcdeabbb'))
                console.log(result[0]);
        })();
```

alexlamsl commented:
Agreed, even though I was hoping to have some test failures when I remove `def(AST_RegExp, return_true);` from `has_side_effects()`. 😅 

FWIW, the corner cases of size regression I've hit can be recovered with that line removed. Which means optimising `AST_RegExp` properly will now be on my TODO list. 👻

alexlamsl commented:
Looking at `AST_Object` & `AST_Function` though:
```js
def(AST_Function, return_false);

def(AST_Object, function(compressor){
    for (var i = this.properties.length; --i >= 0;)
        if (this.properties[i].has_side_effects(compressor))
            return true;
    return false;
});
```
Wouldn't `AST_RegExp` seem more restricted by comparison?

kzc commented:
Technically a `RegExp` literal is shorthand for `new RegExp(...)`. 
```
$ echo 'function f(){var a=new RegExp("b");}' | uglifyjs -c pure_funcs='["RegExp"]'
WARN: Dropping unused variable a [-:1,17]
function f(){}
```
`new RegExp()` or a `RegExp` literal can be considered to be a pure function if otherwise unused. Perhaps the `unused` optimization can codify that specific rule.

alexlamsl commented:
Throwing this at `uglifyjs -c`:
```js
function f() {
  var a = {
    lastIndex: -1,
    flags: "g",
    source: "ab*",
    exec: function(s) {
      return s;
    }
  };
}
```
... yields `function f(){}`. I'm still a bit stuck on why `/ab*/g` is different from this object. 😕

kzc commented:
> I was hoping to have some test failures when I remove def(AST_RegExp, return_true); from has_side_effects().

I suggested that addition just to be on the safe side. `!is_constant()` takes precedence over `has_side_effects()` in `collapse_vars`, but I wasn't sure other `AST_Constant` optimization code would have the same safeguards for `AST_RegExp`.
```
$ uglifyjs -V
uglify-js 2.7.5

$ echo 'function f(){var a=/b/;}' | uglifyjs -c
WARN: Dropping unused variable a [-:1,17]
function f(){}
```
But since 2.7.5 correctly drops the unused RegExp literal, perhaps this PR is too restrictive and `def(AST_RegExp, return_true);` should be dropped from `has_side_effects()`. A test should be added to catch this regression.

kzc commented:
> I'm still a bit stuck on why /ab*/g is different from this object. 

If `has_side_effects()` once again returns `false` for `AST_RegExp` does that solve your issue?

kzc commented:
I'll remove the line in question from #1459

alexlamsl commented:
> If  has_side_effects()  once again returns  false  for  AST_RegExp  does that solve your issue?

Yes, removing that line makes it hit the `AST_Constant` case which returns `false` and recovers back `2.7.5` behaviour.

> I'll remove the line in question from #1459

Thanks - I suppose we can add the tests you suggested over there as well.

kzc commented:
@alexlamsl Regression addressed in #1459 and new test added. Thanks for catching that.

alexlamsl commented:
Right, with two PRs updated, let me run through this merge exercise again... 😎

kzc commented:
@alexlamsl My bad - sorry about that.

alexlamsl commented:
@kzc no worries!

alexlamsl commented:
@kzc with default settings, only `ember.prod.js` is affected by the size regression.

Updated `test/benchmark.js` results. I'll now run through that list of `test/jetstream.js` options.

alexlamsl commented:
All [26 configurations](https://github.com/mishoo/UglifyJS2/pull/1485#issuecomment-279137462) passed `test/jetstream.js` without incident.

kzc commented:
> with default settings, only ember.prod.js is affected by the size regression.

There's an `ember.prod.js` size regression from 2.7.5 with what flags? Can you post the sizes you obtained?

avdg commented:
Mind to merge #1487?

kzc commented:
uglify-js@2.7.5:
```
$ curl http://builds.emberjs.com/tags/v2.11.0/ember.prod.js | uglifyjs -mc warnings=0 | wc -c

  530720
```
This PR:
```
$ curl http://builds.emberjs.com/tags/v2.11.0/ember.prod.js | bin/uglifyjs -mc warnings=0 | wc -c

  530438
```
I don't see a regression with default compress and mangle.

Wait - you meant the `AST_RegExp.has_side_effects()` regression which was resolved. Nevermind!

alexlamsl commented:
> Wait - you meant the  AST_RegExp.has_side_effects()  regression which was resolved.

@kzc yes, I meant that. Sorry for the confusion 😅

alexlamsl commented:
> Mind to merge #1487?

@avdg done 😉

alexlamsl commented:
New conflict, but otherwise #1487 is (re-)merged.

alexlamsl commented:
Rebuilt this PR and removed #1366. Added `closes #<PR>` to each commit so this PR can be used as @kzc suggested in https://github.com/mishoo/UglifyJS2/pull/1485#issuecomment-279160093

alexlamsl commented:
@kzc in merging #1491 I've to modify your #1448 due to it introducing the same bug as what #1491 is trying to fix. I took the liberty to assume if you mark the `AST_Call` as `@__PURE__`, then it take precedence over `options.pure_funcs`.

PTAL at https://github.com/mishoo/UglifyJS2/pull/1485/commits/d07908992eb4cac5857f273c08ec090ca1ae870c

kzc commented:
@alexlamsl Did you mean #1494 rather than #1491? Your assumptions w.r.t. #1448 are fine. It stands to reason it had the same argument side-effect issue as #1494.

Since you're merging this into staging anyway could you copy/paste another test under [this `#__PURE__` test](https://github.com/alexlamsl/UglifyJS2/blob/d07908992eb4cac5857f273c08ec090ca1ae870c/test/compress/issue-1261.js#L1) and enable [`-c toplevel`](https://github.com/mishoo/UglifyJS2/pull/1450) which would result in the elimination of `iife1`? Please update comments w.r.t. `iife1` accordingly in the new test. Thanks!

alexlamsl commented:
> could you copy/paste another test

PTAL https://github.com/mishoo/UglifyJS2/pull/1485/commits/391a25a8ef6cd958ccf3f1b1b86a73147aeb03c5

kzc commented:
@mishoo @rvanvelzen Would you kindly consider granting @alexlamsl github and npm rights to this project to facilitate a release?

alexlamsl commented:
FYI - just had a slow Travis CI test instance and this PR still manages to pass without time-outs.

alexlamsl commented:
`test/jetstream.js` with battery of options from https://github.com/mishoo/UglifyJS2/pull/1496#issuecomment-280853272 passed without incident.

@kzc just a few more sample points for your question in https://github.com/mishoo/UglifyJS2/pull/1496#issuecomment-280858248
```
uglifyjs Octane2/mandreel.js -b beautify=false --stats
Timing information (compressed 1 files):
- parse: 5.634s
- generate: 5.528s

uglifyjs Octane2/mandreel.js -mc unsafe,passes=2,warnings=false --stats
WARN: Output exceeds 32000 characters
Timing information (compressed 1 files):
- parse: 4.731s
- scope: 24.236s
- squeeze: 182.309s
- mangle: 1.220s
- generate: 5.707s

uglifyjs Octane2/mandreel.js -mc reduce_vars,passes=3,warnings=false --stats
WARN: Output exceeds 32000 characters
Timing information (compressed 1 files):
- parse: 4.955s
- scope: 23.538s
- squeeze: 267.733s
- mangle: 1.245s
- generate: 5.557s

uglifyjs Octane2/mandreel.js -mc unsafe,keep_fargs=0,collapse_vars,reduce_vars,warnings=false --stats
WARN: Output exceeds 32000 characters
Timing information (compressed 1 files):
- parse: 5.432s
- scope: 24.052s
- squeeze: 91.206s
- mangle: 1.134s
- generate: 5.340s

uglifyjs Octane2/mandreel.js -mc unsafe,unsafe_comps,pure_getters,keep_fargs=0,collapse_vars,reduce_vars,passes=3,warnings=false --stats
WARN: Output exceeds 32000 characters
Timing information (compressed 1 files):
- parse: 4.950s
- scope: 20.201s
- squeeze: 291.570s
- mangle: 1.063s
- generate: 4.445s
```

kzc commented:
Is there an option in `test/jetstream.js` to save the original JS inputs such as `Octane2/mandreel.js` to `test/jetstream/Octane2/mandreel.js`? Would have to update .gitignore accordingly.

kzc commented:
By AST_Sequence timings I meant this PR specifically:

[WIP][3.x] convert AST_Seq from binary tree to array #1460

It's considerably faster than AST_Seq car/cdr for `test/jetstream.js`

alexlamsl commented:
> By AST_Sequence timings I meant this PR specifically:
> [WIP][3.x] convert AST_Seq from binary tree to array #1460

Which is what I tested and reported in https://github.com/mishoo/UglifyJS2/pull/1496#issuecomment-280861135 - sorry if that wasn't clear enough :sweat_smile:

alexlamsl commented:
> Is there an option in  test/jetstream.js  to save the original JS inputs

Not at the moment. This and reporting before/after/gzip sizes in `test/benchmark.js` are the TODOs I have at the back of my mind after #1479.

kzc commented:
> Which is what I tested and reported in #1496 (comment) - sorry if that wasn't clear enough

@alexlamsl I did see that, thanks. I was just was restating the obvious need for #1460.

kzc commented:
@alexlamsl A fix was made for #1448. Would you be able to fold that change in this PR?

alexlamsl commented:
@kzc #1448 re-merged alongside updated version of https://github.com/mishoo/UglifyJS2/pull/1485/commits/01b8313032d090e3d64250b405a1dc8716cc0cdc

I have moved `has_pure_annotation()` and the associated warning around to facilitate code reuse in `drop_side_effect_free()`.

alexlamsl commented:
@kzc in merging #1498 I've noticed #1459 would suffer the same issue as described in #1497.

I have applied a fix which would make #1459 [dependent on `reduce_vars`](https://github.com/mishoo/UglifyJS2/pull/1485/files#diff-230c53e13e3ee367aec7d49470eb42caR3028) - PTAL.

kzc commented:
@alexlamsl Here's the thing - `const` variable reassignment really ought to throw an error well before we get to the #1459 optimization - otherwise we need to have redundant wasteful checking all over the place. I would not be surprised if there are other optimizations we've missed. But as long as `reduce_vars` is on by default, I can live with it. Your suggestion in https://github.com/mishoo/UglifyJS2/pull/1459#issuecomment-282068834 make sense.

kzc commented:
@alexlamsl Thanks for getting all these features merged into master! 
(I don't envy the person who will merge this into harmony ;-)

Now all you need is npm rights to create a new release.

/cc @mishoo @rvanvelzen
